ignite/pkg/protoanalysis/protoutil/creator.go

```diff
@@ -0,0 +1,651 @@
+	}
+}
+
+// NewField creates a new field statement node:
+//
+//	// int32 Foo = 1;
+//	field := NewField("Foo", "int32", 1)
+//
+// Fields aren't marked as repeated, required or optional. Use Repeated, Optional
+// and Required to mark the field as such.
+//
+//	// repeated int32 Foo = 1;
+//	field := NewField("Foo", "int32", 1, Repeated())
+func NewField(typ, name string, sequence int, opts ...FieldSpecOpts) *proto.NormalField {
+	f := FieldSpec{name: name, typ: typ, sequence: sequence}
+	for _, opt := range opts {
+		opt(&f)
+	}
+
+	// Check qualifiers? Though protoc will shout if we do stupid things.
+	field := &proto.NormalField{
+		Field: &proto.Field{
+			Name:     f.name,
+			Sequence: f.sequence,
+			Type:     f.typ,
+			Options:  []*proto.Option{},
+		},
+		Repeated: f.repeated,
+		Required: f.required,
+		Optional: f.optional,
+	}
+	if len(f.options) > 0 {
+		field.Options = append(field.Options, f.options...)
+	}
+	return field
+}
+
+// MessageSpec holds information relevant to the message statement.
+type MessageSpec struct {
+	name     string
+	fields   []*proto.NormalField // needs a good amount of work.
+	enums    []*proto.Enum
+	options  []*proto.Option
+	isExtend bool
+}
+
+// MessageSpecOpts is a type alias for a callable accepting a MessageSpec.
+type MessageSpecOpts func(i *MessageSpec)
+
+// WithMessageOptions adds options to the message.
+func WithMessageOptions(options ...*proto.Option) MessageSpecOpts {
+	return func(m *MessageSpec) {
+		m.options = append(m.options, options...)
+	}
+}
+
+// WithMessageFields adds fields to the message.
+func WithFields(fields ...*proto.NormalField) MessageSpecOpts {
+	return func(m *MessageSpec) {
+		m.fields = append(m.fields, fields...)
+	}
+}
+
+// WithEnums adds enums to the message.
+func WithEnums(enum ...*proto.Enum) MessageSpecOpts {
+	return func(m *MessageSpec) {
+		m.enums = append(m.enums, enum...)
+	}
+}
+
+func Extend() MessageSpecOpts {
+	return func(m *MessageSpec) {
+		m.isExtend = true
+	}
+}
+
+// NewMessage creates a new message statement node:
+//
+//	// message Foo {}
+//	message := NewMessage("Foo")
+//
+// No fields/enums/options are attached by default, use WithMessageFields, WithEnums,
+// and WithMessageOptions to add them as required:
+//
+//	 // message Foo {
+//	 //  option (foo) = 1;
+//	 //  int32 Bar = 1;
+//	 // }
+//		opt := NewOption("foo", "1")
+//	 field := NewField("int32", "Bar", 1)
+//	 message := NewMessage("Foo", WithMessageOptions(opt), WithFields(field))
+//
+// By default, options are added first, then fields and then enums.
+func NewMessage(name string, opts ...MessageSpecOpts) *proto.Message {
+	m := MessageSpec{name: name}
+	for _, opt := range opts {
+		opt(&m)
+	}
+	message := &proto.Message{
+		Name:     m.name,
+		IsExtend: m.isExtend,
+	}
+	for _, opt := range m.options {
+		message.Elements = append(message.Elements, opt)
+	}
+
+	// Verify that fields have unique sequence? Though, again, protoc will shout if
```

Comment:
Can users directly specify the field numbers when scaffolding through the cli? I don't know how applicable this comment really is since I think we create the sequence via a loop, irrespective of what the user asks for.